src/test/java/lambda/part2/exercise/Exercise2.java

```diff
@@ -0,0 +1,98 @@
+    }
+
+    // метод (Person -> boolean, Person -> boolean) -> (Person -> boolean)
+    // - возвращает новый предикат, объединяющий исходные с помощью операции "AND"
+    // - при реализации использовать логический оператор &&
+    private Predicate<Person> andUsingLogicalOperator(Predicate<Person> left, Predicate<Person> right) {
+        return person -> left.test(person) && right.test(person);
+    }
+
+    @Test
+    void personHasNotEmptyLastNameAndFirstNameUsingLogicalOperators() {
+        Predicate<Person> personHasEmptyFirstName = person -> person.getFirstName().isEmpty();
+        Predicate<Person> personHasEmptyLastName = person -> person.getLastName().isEmpty();
+
+        Predicate<Person> personHasNotEmptyFirstName = negateUsingLogicalOperator(personHasEmptyFirstName);
+        Predicate<Person> personHasNotEmptyLastName = negateUsingLogicalOperator(personHasEmptyLastName);
+
+        Predicate<Person> personHasNotEmptyLastNameAndFirstName = negateUsingLogicalOperator(andUsingLogicalOperator(personHasEmptyLastName, personHasEmptyFirstName));
+
+        assertThat(personHasNotEmptyLastNameAndFirstName.test(new Person("Алексей", "Доренко", 40)), is(true));
+        assertThat(personHasNotEmptyLastNameAndFirstName.test(new Person("Николай", "", 30)), is(true));
+        assertThat(personHasNotEmptyLastNameAndFirstName.test(new Person("", "Мельников", 20)), is(true));
+    }
+
+    // метод (T -> boolean) -> (T -> boolean)
+    // - возвращает новый предикат, являющийся отрицанием исходного
+    // - при реализации использовать логический оператор !
+    private <T> Predicate<T> negate(Predicate<T> predicate) {
+        return t -> !predicate.test(t);
+    }
+
+    // метод (T -> boolean, T -> boolean) -> (T -> boolean)
+    // - возвращает новый предикат, объединяющий исходные с помощью операции "AND"
+    // - при реализации использовать логический оператор &&
+    private <T> Predicate<T> and(Predicate<T> left, Predicate<T> right) {
+        return t -> left.test(t) && right.test(t);
+    }
+
+    @Test
+    void personHasNotEmptyLastNameAndFirstNameUsingGenericPredicates() {
+        Predicate<Person> personHasEmptyFirstName = person -> person.getFirstName().isEmpty();
+        Predicate<Person> personHasEmptyLastName = person -> person.getLastName().isEmpty();
+
+        Predicate<Person> personHasNotEmptyFirstName = negate(personHasEmptyFirstName);
+        Predicate<Person> personHasNotEmptyLastName = negate(personHasEmptyLastName);
+
+        Predicate<Person> personHasNotEmptyLastNameAndFirstName = negate(and(personHasEmptyLastName, personHasEmptyFirstName));
+
+        assertThat(personHasNotEmptyLastNameAndFirstName.test(new Person("Алексей", "Доренко", 40)), is(true));
+        assertThat(personHasNotEmptyLastNameAndFirstName.test(new Person("Николай", "", 30)), is(true));
+        assertThat(personHasNotEmptyLastNameAndFirstName.test(new Person("", "Мельников", 20)), is(true));
```

Comment:
Two tests above should check for false, so if your tests pass something wrong here